hazelcast/src/main/java/com/hazelcast/client/impl/ClientHeartbeatMonitor.java

```diff
@@ -68,13 +73,41 @@ public void start() {
 
     @Override
     public void run() {
+        cleanupEndpointsWithDeathConnections();
+
         String memberUuid = clientEngine.getThisUuid();
         for (ClientEndpoint ce : clientEndpointManager.getEndpoints()) {
             ClientEndpointImpl clientEndpoint = (ClientEndpointImpl) ce;
             monitor(memberUuid, clientEndpoint);
         }
     }
 
+    private void cleanupEndpointsWithDeathConnections() {
+        ArrayList<ClientEndpoint> endpointsWithDeadConnections = new ArrayList<ClientEndpoint>();
```

Comment:
You are going to create litter every time you do a check. There is no need for that. You can iterate over the set of endpoints and remove the dead endpoint as soon as you see it.

Reply:
I can't use the iterate approach since map could be modified when I am iterating over. 
That is why I avoid iterators. 

I can avoid litter by creating the list only if there is a closed connection.